dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/AbstractTrackerValidationHook.java

```diff
@@ -0,0 +1,274 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Objects;
+
+import static org.hisp.dhis.tracker.report.ValidationErrorReporter.newReport;
+import static org.hisp.dhis.tracker.validation.hooks.Constants.PROGRAM_INSTANCE_CANT_BE_NULL;
+import static org.hisp.dhis.tracker.validation.hooks.Constants.TRACKED_ENTITY_ATTRIBUTE_CANT_BE_NULL;
+import static org.hisp.dhis.tracker.validation.hooks.Constants.TRACKED_ENTITY_CANT_BE_NULL;
+import static org.hisp.dhis.tracker.validation.hooks.Constants.TRACKER_BUNDLE_CANT_BE_NULL;
+import static org.hisp.dhis.tracker.validation.hooks.Constants.USER_CANT_BE_NULL;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+public abstract class AbstractTrackerValidationHook
+    implements TrackerValidationHook
+
+{
+    @Autowired
+    protected TrackedEntityAttributeService teAttrService;
+
+    @Autowired
+    protected TrackerAccessManager trackerAccessManager;
+
+    @Autowired
+    protected OrganisationUnitService organisationUnitService;
+
+    @Autowired
+    protected ProgramInstanceService programInstanceService;
+
+    @Override
+    public boolean isEnabled()
+    {
+        return true;
+    }
+
+    protected void validateGeometryFromCoordinates( ValidationErrorReporter errorReporter, String coordinates,
+        FeatureType featureType )
+    {
+        Objects.requireNonNull( featureType, "FeatureType can't be null" );
+
+        if ( coordinates != null && FeatureType.NONE != featureType )
+        {
+            try
+            {
+                GeoUtils.getGeometryFromCoordinatesAndType( featureType, coordinates );
+            }
+            catch ( IOException e )
+            {
+                errorReporter.addError( newReport( TrackerErrorCode.E1013 )
+                    .addArg( coordinates )
+                    .addArg( e.getMessage() ) );
+            }
+        }
+    }
+
+    protected void validateAttrValueType( ValidationErrorReporter errorReporter, Attribute attr,
+        TrackedEntityAttribute teAttr )
+    {
+        Objects.requireNonNull( attr, Constants.ATTRIBUTE_CANT_BE_NULL );
+        Objects.requireNonNull( teAttr, TRACKED_ENTITY_ATTRIBUTE_CANT_BE_NULL );
+
+        String error = teAttrService.validateValueType( teAttr, attr.getValue() );
+        if ( error != null )
+        {
+            errorReporter.addError( newReport( TrackerErrorCode.E1007 )
+                .addArg( error ) );
+        }
+    }
+
+    protected void validateAttributeUniqueness( ValidationErrorReporter errorReporter,
+        String value,
+        TrackedEntityAttribute trackedEntityAttribute,
+        TrackedEntityInstance trackedEntityInstanceUid,
+        OrganisationUnit organisationUnit )
+    {
+        Objects.requireNonNull( trackedEntityAttribute, TRACKED_ENTITY_ATTRIBUTE_CANT_BE_NULL );
+
+        if ( Boolean.TRUE.equals( trackedEntityAttribute.isUnique() ) )
```

Comment:
Skip out if false, no need for extra indent level (unless you plan to add extra code here)